### PR TITLE
[refactor]: split `iroha_torii` from `iroha`

### PR DESCRIPTION
**Pull request by 0x009922**
## Description

This PR splits Torii (web server functionality), located in the main CLI, into a separate `iroha_torii` crate.

The rationale is to make a step towards making the main CLI exclusively a binary. Currently `iroha` is a library as well, exposing functionality to run `Iroha` in tests. That includes Torii. Since Torii is quite self-contained, making it a separate crate seem to make sense.

A further step would be to move `Iroha` itself and `iroha::samples` out of the CLI, so that `test_network` doesn't depend on `iroha` anymore, and `iroha` might be exclusively a binary.

Additionally:

- fix `telemetry` feature flag. Torii code didn't compile if this feature was disabled.
- clean dependencies of `iroha` CLI

### Linked issue

Addresses #4136, but doesn't close it.

### CODE REVIEW

**Comment by mversic** [comment]
I'm not sure we need this. I'd like to get others opinions on this. Having multiple crates might speed up compilation, on the other hand `lib` and `bin` should be compiled as separate crates

**Comment by 6r1d** [comment]
Given the [rationale](https://github.com/hyperledger/iroha/pull/4139#issue-2037569803), I don't see the [compilation speed](https://github.com/hyperledger/iroha/pull/4139#issuecomment-1853426909) as a main aspect. I agree with the rationale, and I'm in favour of the change.

**Review by mversic — APPROVED** [review]

File: cli/Cargo.toml [hunk 1]
```diff
@@ -91,10 +76,10 @@ vergen = { workspace = true, features = ["cargo"] }
 
 [package.metadata.cargo-all-features]
 denylist = [
-"bridge",
-"dex",
-"schema-endpoint",
-"telemetry",
-"test-network"
+  "bridge",
+  "dex",
```
Comment (Arjentix) [inline cli/Cargo.toml:80]: Oh, we should remove that confusing stuff someday

File: torii/derive/src/lib.rs
Comment (Arjentix) [inline torii/derive/src/lib.rs:1]: Well, there are no derives here actually. Maybe rename to `macro`?

**Review by Arjentix — APPROVED** [review]

File: cli/Cargo.toml (hunk 1 above)
Comment (0x009922) [reply cli/Cargo.toml:80]: Issued https://github.com/hyperledger/iroha/issues/4152

File: torii/derive/src/lib.rs
Comment (0x009922) [reply torii/derive/src/lib.rs:1]: Done.

Comment (mversic) [reply torii/derive/src/lib.rs:1]: should we do this consistently across codebase? I thought the practice was to name proc-macro crates `derive`. Whether or not they had `derive` macros, `attribute` macros or `function-like` macros

**Comment by Erigara** [comment]
I just get another idea why using splitting torii might be good idea. 
I think for some network setups it might be useful to have nodes for fault tolerance, but this nodes want serve users requests and will only participate in consensus.

File: torii/src/routing.rs [hunk 2]
```diff
@@ -179,7 +180,7 @@ async fn handle_post_configuration(
 }
 
 #[iroha_futures::telemetry_future]
-async fn handle_blocks_stream(kura: Arc<Kura>, mut stream: WebSocket) -> eyre::Result<()> {
+pub async fn handle_blocks_stream(kura: Arc<Kura>, mut stream: WebSocket) -> eyre::Result<()> {
```
Comment (Erigara) [inline torii/src/routing.rs:183]: Why this functions have to be public now?

File: torii/derive/src/lib.rs
Comment (Erigara) [reply torii/derive/src/lib.rs:1]: I also thought that it's a convention, however i'm not sure about this anymore (e.g. `tokio` calls it's proc macro crate `tokio-macros`)

File: torii/src/routing.rs (hunk 2 above)
Comment (0x009922) [reply torii/src/routing.rs:183]: Because `Torii::new` now defined next to `Torii` itself, in the parent module.

**Comment by 0x009922** [comment]
> So torii could be excluded from compilation.

Well, Torii might be excluded with feature flags anyway, without necessarily being a separate crate.

**Review by Erigara — APPROVED** [review]

File: torii/derive/src/lib.rs
Comment (0x009922) [reply torii/derive/src/lib.rs:1]: Created https://github.com/hyperledger/iroha/issues/4154

**Review by Arjentix — APPROVED** [review]

**Pull request merged**